crates/red_knot_module_resolver/src/path.rs

```diff
@@ -0,0 +1,693 @@
+            }
+            ModuleResolutionPath::StandardLibrary(StandardLibraryPathBuf(path)) => {
+                ModuleResolutionPathRef::standard_library_unchecked(path)
+            }
+            ModuleResolutionPath::SitePackages(SitePackagesPathBuf(path)) => {
+                ModuleResolutionPathRef::site_packages_unchecked(path)
+            }
+        }
+    }
+}
+
+impl<'a> AsRef<FileSystemPath> for ModuleResolutionPathRef<'a> {
+    #[inline]
+    fn as_ref(&self) -> &FileSystemPath {
+        self.as_file_system_path()
+    }
+}
+
+impl<'a> PartialEq<ModuleResolutionPath> for ModuleResolutionPathRef<'a> {
+    fn eq(&self, other: &ModuleResolutionPath) -> bool {
+        match (self, other) {
+            (
+                ModuleResolutionPathRef::Extra(ExtraPath(self_path)),
+                ModuleResolutionPath::Extra(ExtraPathBuf(other_path)),
+            )
+            | (
+                ModuleResolutionPathRef::FirstParty(FirstPartyPath(self_path)),
+                ModuleResolutionPath::FirstParty(FirstPartyPathBuf(other_path)),
+            )
+            | (
+                ModuleResolutionPathRef::StandardLibrary(StandardLibraryPath(self_path)),
+                ModuleResolutionPath::StandardLibrary(StandardLibraryPathBuf(other_path)),
+            )
+            | (
+                ModuleResolutionPathRef::SitePackages(SitePackagesPath(self_path)),
+                ModuleResolutionPath::SitePackages(SitePackagesPathBuf(other_path)),
+            ) => *self_path == **other_path,
+            _ => false,
+        }
+    }
+}
+
+impl<'a> PartialEq<ModuleResolutionPathRef<'a>> for ModuleResolutionPath {
+    fn eq(&self, other: &ModuleResolutionPathRef<'a>) -> bool {
+        other.eq(self)
+    }
+}
+
+impl<'a> PartialEq<FileSystemPath> for ModuleResolutionPathRef<'a> {
+    fn eq(&self, other: &FileSystemPath) -> bool {
+        self.as_file_system_path() == other
+    }
+}
+
+impl<'a> PartialEq<ModuleResolutionPathRef<'a>> for FileSystemPath {
+    fn eq(&self, other: &ModuleResolutionPathRef<'a>) -> bool {
+        self == other.as_file_system_path()
+    }
+}
+
+impl<'a> PartialEq<VfsPath> for ModuleResolutionPathRef<'a> {
+    fn eq(&self, other: &VfsPath) -> bool {
+        let VfsPath::FileSystem(other) = other else {
+            return false;
+        };
+        self.as_file_system_path() == &**other
+    }
+}
+
+impl<'a> PartialEq<ModuleResolutionPathRef<'a>> for VfsPath {
+    fn eq(&self, other: &ModuleResolutionPathRef<'a>) -> bool {
+        other.eq(self)
+    }
+}
+
+pub(crate) struct ModulePartIterator<'a> {
+    parent_components: Option<camino::Utf8Components<'a>>,
+    stem: Option<&'a str>,
+}
+
+impl<'a> ModulePartIterator<'a> {
+    #[must_use]
+    fn from_fs_path(path: &'a FileSystemPath) -> Self {
+        Self {
+            parent_components: path.parent().map(|path| path.components()),
+            stem: path.file_stem(),
+        }
+    }
+}
```

Comment:
Nit: I think you can avoid the option here
```suggestion
    parent_components: Option<camino::Utf8Components<'a>>,
    stem: Option<&'a str>,
}

impl<'a> ModulePartIterator<'a> {
    #[must_use]
    fn from_fs_path(path: &'a FileSystemPath) -> Self {
        let mut parent_components = path.components();
        parent_components.next_back();
        Self {
            parent_components,
            stem: path.file_stem(),
        }
    }
}
```

This should also simplify the `next` implementation

Reply:
Hmm, not sure if that works or not. Let me write some tests for this bit of code before applying this change.

Reply:
You're right, this does do the right thing. The thing I was confused about is for some reason I wasn't sure whether `next_back()` calls consumed the end of the iterator in a way that would then be reflected when walking the iterator from the start. But it makes sense that it would be.